File: Assets/Typescript~/src/Shared/UserInput/Keyboard.ts
import { Bin } from "Shared/Util/Bin";
import { Signal } from "Shared/Util/Signal";
import { SignalProxy } from "Shared/Util/SignalProxy";
import { KeyboardDriver } from "./Drivers/KeyboardDriver";
import { KeySignal } from "./Drivers/Signals/KeySignal";

export class Keyboard {
	private readonly bin = new Bin();
	private readonly keyboardDriver = KeyboardDriver.instance();
	private readonly keysDown = new Map<Key, boolean>();

	/** The `KeyDown` signal fires any time a key is pressed down. */
	public readonly KeyDown: Signal<[event: KeySignal]>;

	/** The `KeyUp` signal fires any time a key is released. */
	public readonly KeyUp = new Signal<KeySignal>();

	constructor() {
		this.KeyDown = this.bin.Add(new SignalProxy<[KeySignal]>(this.keyboardDriver.KeyDown));
		this.bin.Add(this.KeyDown);
		this.bin.Add(this.KeyUp);
		this.bin.Connect(this.keyboardDriver.KeyDown, (key) => {
			this.keysDown.set(key.Key, true);
			this.KeyDown.Fire(key);
		});
		this.bin.Connect(this.keyboardDriver.KeyUp, (key) => {
			this.keysDown.set(key.Key, false);
			this.KeyUp.Fire(key);
		});
	}

	/** Returns `true` if the given key is down. */
	public IsKeyDown(key: Key) {
		let keyDown = this.keysDown.get(key);
		if (keyDown === undefined) {
			keyDown = this.keyboardDriver.IsKeyDown(key);
			this.keysDown.set(key, keyDown);
		}
		return keyDown;
	}

	/** Returns `true` if either of the given keys are down. */
	public IsEitherKeyDown(key1: Key, key2: Key) {
		return this.IsKeyDown(key1) || this.IsKeyDown(key2);
	}

	/** Returns `true` if both keys are down. */
	public AreBothKeysDown(key1: Key, key2: Key) {
		return this.IsKeyDown(key1) && this.IsKeyDown(key2);
	}

	public Destroy() {
		this.bin.Destroy();
	}
}

File: Assets/Bundles/Shared/Resources/TS/UserInput/Keyboard.lua
-- Compiled with unity-ts v2.1.0-75
local Bin = require("Shared/TS/Util/Bin").Bin
local Signal = require("Shared/TS/Util/Signal").Signal
local SignalProxy = require("Shared/TS/Util/SignalProxy").SignalProxy
local KeyboardDriver = require("Shared/TS/UserInput/Drivers/KeyboardDriver").KeyboardDriver
local Keyboard
do
	Keyboard = setmetatable({}, {
		__tostring = function()
			return "Keyboard"
		end,
	})
	Keyboard.__index = Keyboard
	function Keyboard.new(...)
		local self = setmetatable({}, Keyboard)
		return self:constructor(...) or self
	end
	function Keyboard:constructor()
		self.bin = Bin.new()
		self.keyboardDriver = KeyboardDriver:instance()
		self.keysDown = {}
		self.KeyUp = Signal.new()
		self.KeyDown = self.bin:Add(SignalProxy.new(self.keyboardDriver.KeyDown))
		self.bin:Add(self.KeyDown)
		self.bin:Add(self.KeyUp)
		self.bin:Connect(self.keyboardDriver.KeyDown, function(key)
			local _keysDown = self.keysDown
			local _key = key.Key
			_keysDown[_key] = true
			self.KeyDown:Fire(key)
		end)
		self.bin:Connect(self.keyboardDriver.KeyUp, function(key)
			local _keysDown = self.keysDown
			local _key = key.Key
			_keysDown[_key] = false
			self.KeyUp:Fire(key)
		end)
	end
	function Keyboard:IsKeyDown(key)
		local _keysDown = self.keysDown
		local _key = key
		local keyDown = _keysDown[_key]
		if keyDown == nil then
			keyDown = self.keyboardDriver:IsKeyDown(key)
			local _keysDown_1 = self.keysDown
			local _key_1 = key
			local _keyDown = keyDown
			_keysDown_1[_key_1] = _keyDown
		end
		return keyDown
	end
	function Keyboard:IsEitherKeyDown(key1, key2)
		return self:IsKeyDown(key1) or self:IsKeyDown(key2)
	end
	function Keyboard:AreBothKeysDown(key1, key2)
		return self:IsKeyDown(key1) and self:IsKeyDown(key2)
	end
	function Keyboard:Destroy()
		self.bin:Destroy()
	end
end
return {
	Keyboard = Keyboard,
}
-- ----------------------------------
-- ----------------------------------
